dart/lib/vm_service_lib.dart

```diff
@@ -1427,18 +1474,57 @@ class Class extends Obj {
 
   Class._fromJson(Map<String, dynamic> json) : super._fromJson(json) {
     name = json['name'];
-    error = _createObject(json['error']);
+    error = createServiceObject(json['error']);
     isAbstract = json['abstract'];
     isConst = json['const'];
-    library = _createObject(json['library']);
-    location = _createObject(json['location']);
-    superClass = _createObject(json['super']);
-    superType = _createObject(json['superType']);
-    interfaces = new List<InstanceRef>.from(_createObject(json['interfaces']));
-    mixin = _createObject(json['mixin']);
-    fields = new List<FieldRef>.from(_createObject(json['fields']));
-    functions = new List<FuncRef>.from(_createObject(json['functions']));
-    subclasses = new List<ClassRef>.from(_createObject(json['subclasses']));
+    library = createServiceObject(json['library']);
+    location = createServiceObject(json['location']);
+    superClass = createServiceObject(json['super']);
+    superType = createServiceObject(json['superType']);
+    interfaces =
+        new List<InstanceRef>.from(createServiceObject(json['interfaces']));
+    mixin = createServiceObject(json['mixin']);
+    fields = new List<FieldRef>.from(createServiceObject(json['fields']));
+    functions = new List<FuncRef>.from(createServiceObject(json['functions']));
+    subclasses =
+        new List<ClassRef>.from(createServiceObject(json['subclasses']));
+  }
+
+  Map<String, dynamic> toJson() {
+    var json = super.toJson();
+    json['type'] = 'Class';
+    json.addAll({
+      'name': name,
+      'abstract': isAbstract,
+      'const': isConst,
+      'library': library.toJson(),
+      'interfaces': interfaces.map((f) => f.toJson()).toList(),
+      'fields': fields.map((f) => f.toJson()).toList(),
+      'functions': functions.map((f) => f.toJson()).toList(),
+      'subclasses': subclasses.map((f) => f.toJson()).toList(),
+    });
+    var nextVal;
+    nextVal = error?.toJson();
```

Comment:
Note that I am not super happy with this pattern, so I am definitely open to suggestions. The reasons I did it this way:

- don't want to duplicate the work of evaluating the expression in the if and assignment (could be arbitrarily expensive)
- avoid creating a new local variable for each option (although now we have dynamic dispatch always, so I could change this if desired)

One option we could do:

- Create a helper method that takes a map (`json` in this case), a key (to set), and a value (we would inline the expression). It would only set the key on the map if the value isn't null.

Reply:
Example of what this would look like with the helper method:

```dart
void _setIfNotNull(Map<String, Object> json, String key, Object value) {
  if (value == null) return;
  json[key] = value;
}
// later in some toJson
Map<String, Object> toJson() {
   var json = <String, Object>{};
   _setIfNotNull(json, 'error', error?.toJson());
   _setIfNotNull(json, 'location', location?.toJson());
   _setIfNotNull(json, 'super', superClass?.toJson());
   _setIfNotNull(json, 'superType', superType?.toJson());
   _setIfNotNull(json, 'mixin', mixin?.toJson());
}
```

Reply:
Actually I think this is clearly better and closer to what I would write by hand, so I will go ahead and push a commit to change the codegen to look like this.